alacritty/src/display/mod.rs

```diff
@@ -1366,55 +1313,316 @@ fn window_size(
     PhysicalSize::new(width as u32, height as u32)
 }
 
+/// The action performed by [`StrShortener`].
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum TextAction {
+    /// Yeild a spacer.
+    Spacer,
+    /// Terminate state reached.
+    Terminate,
+    /// Yeild a shortener.
+    Shortener,
+    /// Yield a character.
+    Char,
+}
+
+/// The direction which we should shorten.
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum ShortenDirection {
+    /// Shorten to the start of the string.
+    Left,
+
+    /// Shorten to the end of the string.
+    Right,
+}
+
+/// Iterator that yeild shortened version of the text.
+struct StrShortener<'a> {
+    chars: Skip<Chars<'a>>,
+    accumulted_len: usize,
+    max_width: usize,
+    direction: ShortenDirection,
+    shortener: Option<char>,
+    text_action: TextAction,
+}
+
+impl<'a> StrShortener<'a> {
+    pub fn new(
+        text: &'a str,
+        max_width: usize,
+        direction: ShortenDirection,
+        mut shortener: Option<char>,
+    ) -> Self {
+        if text.is_empty() {
+            // If we don't have any text don't produce a shortener for it.
+            let _ = shortener.take();
+        }
+
+        let (skip_chars, text_action) = match direction {
+            ShortenDirection::Right => (0, TextAction::Char),
+            ShortenDirection::Left => {
```

Comment:
Might be easier to early-return for ShortenDirection::Right here? I'm not entirely sure, but this is quite a massive block with two easily avoidable indentations.